contracts/governance/rewards/RewardsDistributor.sol

```diff
@@ -0,0 +1,141 @@
+        return MerkleProof.verify(merkleProof, merkleData.root, node);
+    }
+
+    /// @notice Claim accumulated rewards for a set of tokens at a given cycle number
+    function claim(
+        uint256 cycle,
+        uint256 index,
+        IERC20Ext[] calldata tokens,
+        uint256[] calldata cumulativeAmounts,
+        bytes32[] calldata merkleProof
+    ) external nonReentrant returns (uint256[] memory claimAmounts) {
+        require(cycle == merkleData.cycle, 'incorrect cycle');
+
+        // verify the merkle proof
+        bytes32 node = keccak256(abi.encodePacked(cycle, index, msg.sender, tokens, cumulativeAmounts));
+        require(MerkleProof.verify(merkleProof, merkleData.root, node), 'invalid proof');
+
+        claimAmounts = new uint256[](tokens.length);
+
+        // claim each token
+        for (uint256 i = 0; i < tokens.length; i++) {
+            // if none claimable, skip
+            if (cumulativeAmounts[i] == 0) continue;
+
+            uint256 claimable = cumulativeAmounts[i].sub(claimedAmounts[msg.sender][tokens[i]]);
+            if (claimable == 0) continue;
+
+            claimedAmounts[msg.sender][tokens[i]] = cumulativeAmounts[i];
+            claimAmounts[i] = claimable;
+            if (tokens[i] == ETH_TOKEN_ADDRESS) {
+                (bool success, ) = msg.sender.call{ value: claimable }('');
+                require(success, 'eth transfer failed');
+            } else {
+                tokens[i].safeTransfer(msg.sender, claimable);
+            }
+            emit Claimed(cycle, msg.sender, tokens[i], claimable);
+        }
+    }
+
+    // @notice Propose a new root and content hash, only by admin
+    function proposeRoot(
+        uint256 cycle,
+        bytes32 root,
+        string calldata contentHash
+    ) external onlyAdmin {
+        require(cycle == merkleData.cycle.add(1), 'incorrect cycle');
+
+        merkleData.cycle = cycle;
+        merkleData.root = root;
+        merkleData.contentHash = contentHash;
+
+        emit RootUpdated(cycle, root, contentHash);
+    }
+
+    function updateTreasuryPool(IPool _treasuryPool) external onlyAdmin {
+        require(_treasuryPool != IPool(0), 'invalid treasury pool');
+        treasuryPool = _treasuryPool;
+        emit TreasuryPoolSet(_treasuryPool);
+    }
+  
+    function pullFundsFromTreasury(
+        IERC20Ext[] calldata tokens,
+        uint256[] calldata amounts
+    ) external onlyAdmin {
+        treasuryPool.withdrawFunds(tokens, amounts, payable(address(this)));
+    }
+
+    function getClaimedAmounts(
+        address user,
+        IERC20Ext[] calldata tokens
+    ) public view returns (uint256[] memory userClaimedAmounts) {
+        userClaimedAmounts = new uint256[](tokens.length);
+        for (uint256 i = 0; i < tokens.length; i++) {
+            userClaimedAmounts[i] = claimedAmounts[user][tokens[i]];
+        }
+    }
+
+    function encodeClaim(
```

Comment:
Not sure, when this function is used

Reply:
changed to external... it's not really needed actually.. simply copied over from BadgerTree contract